trust-quorum/gfss/src/gf256.rs

```diff
@@ -0,0 +1,302 @@
+}
+impl Add for Gf256 {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        Gf256(self.0 ^ rhs.0)
+    }
+}
+
+impl Add<&Gf256> for Gf256 {
+    type Output = Self;
+    fn add(self, rhs: &Gf256) -> Self::Output {
+        self + *rhs
+    }
+}
+
+impl AddAssign for Gf256 {
+    fn add_assign(&mut self, rhs: Self) {
+        *self = *self + rhs
+    }
+}
+
+impl MulAssign for Gf256 {
+    fn mul_assign(&mut self, rhs: Self) {
+        *self = *self * rhs
+    }
+}
+
+impl Sub for Gf256 {
+    type Output = Self;
+
+    fn sub(self, rhs: Self) -> Self {
+        Gf256(self.0 ^ rhs.0)
+    }
+}
+
+impl Div for Gf256 {
+    type Output = Self;
+
+    fn div(self, rhs: Self) -> Self::Output {
+        self * rhs.invert()
+    }
+}
+
+/// This is an efficient carryless multiplication + reduction algorithm.
+///
+/// It is essentially long multiplication of polynomial elements in GF(2^8)
+/// represented as unsigned bytes (u8). For every bit set in `rhs` we multiply
+/// our original value by `x^i` where i is the bit position of `rhs` and `x` is
+/// our placeholder variable for GF(2^8) operations, and then add the result to
+/// an accumulator initialized to zero. If our accumulator exceeds 255, meaning
+/// it is no longer inside the finite field, we reduce it modulo our irreducible
+/// (prime) polynomial:
+///
+///    m(x) = x^8 + x^4 + x^3 + x + 1
+///
+/// As an example, with our accumulator named `product`:
+///
+///    product = 0
+///    self = 0x21 = 0b0010_0001 = x^5 + 1
+///    rhs  = 0x12 = 0b0001_0010 = x^4 + x
+///
+///    step 1 = ((x^5 + 1) * x) mod m(x)
+///           = x^6 + x
+///    
+///    product += step1
+///
+///    step 2 = ((x^5 + 1) * x^4) mod m(x)
+///           = (x^9 + x^4) mod (x^8 + x^4 + x^3 +x + 1)
+///           = x^5 + x^2 + x
+///
+///    product += step2
+///            = x^6 + x + x^5 + x^2 + x
+///            = x^6 + x^5 + x^2
+///            = 0b0110_0100
+///            = 0x64
+///
+/// See the `test_docs_example` unit test at the bottom of this file to confirm
+/// this math.
+///   
+/// In this algorithm, we use the value `0x1b`, which is our irreducible
+/// polynomial without the high term. We do this because of the following
+/// equality in GF(2^8), which follows from long-division of polynomials:
+///
+///    `x^8 mod m(x) = x^4 + x^3 + x + 1 = 0x1b`
+///
+/// A rationale and description of this algorithm can be found in sections 7.9
+/// and 7.10 of the lecture notes at
+/// <https://engineering.purdue.edu/kak/compsec/NewLectures/Lecture7.pdf>
+///
+/// It is also explained on a wikipedia section about the
+/// [AES finite field](https://en.wikipedia.org/wiki/Finite_field_arithmetic#Rijndael's_(AES)_finite_field)
+///
+/// This is roughly the same algorithm used in
+/// [vsss-rs](https://github.com/mikelodder7/vsss-rs) and many other existing
+/// implementations.
+impl Mul for Gf256 {
+    type Output = Self;
+    fn mul(mut self, mut rhs: Self) -> Self::Output {
```

Comment:
these both being `mut` seems strange, especially since the trait doesn't look to have it https://doc.rust-lang.org/stable/core/ops/trait.Mul.html . I know this is rust but are we going to run into surprises with the `mut`?

Reply:
I completely failed the by value part. I think I was too focused on comparing against the other code which does use temporaries.